lite/examples/object_detection/ios/ObjectDetection/ModelDataHandler/ModelDataHandler.swift

```diff
@@ -265,28 +266,56 @@ class ModelDataHandler: NSObject {
     isModelQuantized: Bool
   ) -> Data? {
     CVPixelBufferLockBaseAddress(buffer, .readOnly)
-    defer { CVPixelBufferUnlockBaseAddress(buffer, .readOnly) }
-    guard let mutableRawPointer = CVPixelBufferGetBaseAddress(buffer) else {
+    defer {
+      CVPixelBufferUnlockBaseAddress(buffer, .readOnly)
+    }
+    guard let sourceData = CVPixelBufferGetBaseAddress(buffer) else {
       return nil
     }
-    assert(CVPixelBufferGetPixelFormatType(buffer) == kCVPixelFormatType_32BGRA)
-    let count = CVPixelBufferGetDataSize(buffer)
-    let bufferData = Data(bytesNoCopy: mutableRawPointer, count: count, deallocator: .none)
-    var rgbBytes = [UInt8](repeating: 0, count: byteCount)
-    var pixelIndex = 0
-    for component in bufferData.enumerated() {
-      let bgraComponent = component.offset % bgraPixel.channels;
-      let isAlphaComponent = bgraComponent == bgraPixel.alphaComponent;
-      guard !isAlphaComponent else {
-        pixelIndex += 1
-        continue
-      }
-      // Swizzle BGR -> RGB.
-      let rgbIndex = pixelIndex * rgbPixelChannels + (bgraPixel.lastBgrComponent - bgraComponent)
-      rgbBytes[rgbIndex] = component.element
+    
+    let width = CVPixelBufferGetWidth(buffer)
+    let height = CVPixelBufferGetHeight(buffer)
+    let sourceBytesPerRow = CVPixelBufferGetBytesPerRow(buffer)
+    let destinationChannelCount = 3
+    let destinationBytesPerRow = destinationChannelCount * width
+    
+    var sourceBuffer = vImage_Buffer(data: sourceData,
+                                     height: vImagePixelCount(height),
+                                     width: vImagePixelCount(width),
+                                     rowBytes: sourceBytesPerRow)
+    
+    guard let destinationData = malloc(height * destinationBytesPerRow) else {
+      print("Error: out of memory")
+      return nil
+    }
+    
+    defer {
+      free(destinationData)
+    }
+
+    var destinationBuffer = vImage_Buffer(data: destinationData,
+                                          height: vImagePixelCount(height),
+                                          width: vImagePixelCount(width),
+                                          rowBytes: destinationBytesPerRow)
+    
+    if (CVPixelBufferGetPixelFormatType(buffer) == kCVPixelFormatType_32BGRA){
+      vImageConvert_BGRA8888toRGB888(&sourceBuffer, &destinationBuffer, UInt32(kvImageNoFlags))
+    } else if (CVPixelBufferGetPixelFormatType(buffer) == kCVPixelFormatType_32ARGB) {
+      vImageConvert_ARGB8888toRGB888(&sourceBuffer, &destinationBuffer, UInt32(kvImageNoFlags))
+    }
+
+    let byteData = Data(bytes: destinationBuffer.data, count: destinationBuffer.rowBytes * height)
+    if isModelQuantized {
+      return byteData
+    }
+
+    // Not quantized, convert to floats
+    let bytes = Array<UInt8>(unsafeData: byteData)!
+    var floats = [Float](repeating: 0, count: bytes.count)
```

Comment:
Is this better/faster than using the existing:

return Data(copyingBufferOf: bytes.map{ Float($0) / 255.0 });

? I'm not super familiar with Swift (and best practices).

Reply:
It was a surprise to me, but it's faster this way. On my 600x600 pixel buffer, the byte to float conversion was 60msec with this code, and 180msec with the map function. It was also slow using `for byte in buffer` enumeration which would be more readable than `for i in 0..<buffer.count`. So my code is faster but I would hesitate to call it "better" because it's unusual and it's not obvious it's written that way for speed.

Reply:
Thanks. Out of curiosity, which device were you testing on?

Reply:
iPhone X